cpp/include/cuspatial/detail/cartesian_product_iterator.cuh

```diff
@@ -0,0 +1,227 @@
+
+    // relative index
+    int32_t relative_idx        = idx - group_block_offset_a - group_block_offset_b;
+    auto relative_element_idx_a = relative_idx % group_size_a;
+    auto relative_element_idx_b = relative_idx / group_size_a;
+
+    return cartesian_product_group_index{{group_idx_a, group_size_a, group_offset_a},
+                                         {group_idx_b, group_size_b, group_offset_b},
+                                         relative_element_idx_a,
+                                         relative_element_idx_b};
+  }
+};
+
+template <typename OffsetIterator>
+struct group_lookup_functor {
+  OffsetIterator group_offsets;
+  int32_t group_count;
+  __device__ int32_t operator()(int32_t element_idx) const
+  {
+    return thrust::distance(
+             group_offsets,
+             thrust::upper_bound(
+               thrust::seq, group_offsets, group_offsets + group_count, element_idx)) -
+           1;
+  }
+};
+
+/**
+ * @brief Makes an iterator which consecutively produces a cartesian product of two iterators.
+ *
+ * Pairs of elements grouped by offsets `A` and `B` are produced consecutively.
+ * - Each group `A_i` in `A` appears consecutively.
+ * - Under each group `A_i`, each group `B_i` in `B` appear consecutively.
+ * - Under each group `B_i`, each element within `a` appears consecutively.
+ * - Under each element within `A_i`, each element within `B_i` appears consecutively.
+ *
+ * Example:
+ * ```
+ * A_elements = [u, v, w, x, y, z]
+ * B_elements = [0, 1, 2, 3, 4, 5]
+ * A_offsets  = [0, 3, 4]
+ * B_offsets  = [0, 2, 5]
+ * ```
+ * Cartesian Product:
+ *
+ *       B0        B1             B2
+ *     +---------+--------------+----+
+ *  A0 : u0   u1 : u2   u3   u4 : u5 :
+ *     +         :              :    +
+ *     : v0   v1 : v2   v3   v4 : v5 :
+ *     +         :              :    +
+ *     : w0   w1 : w2   w3   w4 : w5 :
+ *     +---------+--------------+----+
+ *  A1 : x0   x1 : x2   x3   x4 : x5 :
+ *     +---------+--------------+----+
+ *  A2 : y0   y1 : y2   y3   y4 : y5 :
+ *     +         :              :    +
+ *     : z0   z1 : z2   z3   z4 : z5 :
+ *     +---------+--------------+----+
+ *
+ * Order of iteration:
+ *
+ *  u0, v0, w0, u1, v1, w1, u2, v2, w2, u3, v3, w3, u4, v4, w4, u5, v5, w5, x0, x1, x2, x3, x4, x5,
+ *  y0, z0, y1, z1, y2, z2, y3, z3, y4, z4, y5, z5
+ *
+ * i.e:
+ *
+ *  A0 * B0 | u0 v0 w0 u1 v1 w1
+ *  A0 * B1 | u2 v2 w2 u3 v3 w3 u4 v4 w4
+ *  A0 * B2 | u5 v5 w5
+ *  A1 * B0 | x0 x1
+ *  A1 * B1 | x2 x3 x4
+ *  A1 * B2 | x5
+ *  A1 * B0 | y0 z0 y1 z1
+ *  A2 * B1 | y2 z2 y3 z3 y4 z4
+ *  A2 * B2 | y5 z5
+ *
+ * @tparam OffsetIteratorA
+ * @tparam OffsetIteratorB
+ * @param num_elements_a  number of elements from all groups in A
+ * @param num_elements_b  number of elements from all groups in B
+ * @param num_groups_a    number of groups in A
+ * @param num_groups_b    number of groups in B
+ * @param group_offsets_a offsets for each group in A, plus an "end" offset
```

Comment:
```suggestion
 * @param group_offsets_a Iterator for the offset of the start of each group in A, plus an "end" offset
```

The end offset needs explanation. It's not obvious since this is an iterator, not an array.